File: android/src/main/kotlin/com/cmc/c_shield_sdk/bridges/SslBridge.kt
package com.cmc.c_shield_sdk.bridges

import android.os.Handler
import android.os.Looper
import android.util.Base64
import com.cmc.c_shield_sdk.CShieldErrorCode
import com.example.c_shield_sdk.aip.pinning.CShieldSSL
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import java.io.ByteArrayInputStream
import java.security.cert.CertificateException
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody

class SslBridge {

    // Network I/O must not run on the platform (main) thread; results are posted
    // back to the main thread because MethodChannel.Result must be invoked there.
    private val executor = Executors.newCachedThreadPool()
    private val mainHandler = Handler(Looper.getMainLooper())

    fun handle(call: MethodCall, result: Result) {
        when (call.method) {
            "ssl.configure"          -> handleConfigure(call, result)
            "ssl.updatePins"         -> handleUpdatePins(call, result)
            "ssl.isConfigured"       -> result.success(CShieldSSL.isConfigured())
            "ssl.checkServerTrusted" -> handleCheckServerTrusted(call, result)
            "ssl.httpRequest"        -> handleHttpRequest(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handleConfigure(call: MethodCall, result: Result) {
        val pins = call.argument<List<String>>("pins")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "pins required", null)
        val hostname = call.argument<String>("hostname")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "hostname required", null)
        try {
            CShieldSSL.configure(pins = pins, hostname = hostname)
            result.success(null)
        } catch (e: Exception) {
            result.error(CShieldErrorCode.NATIVE_ERROR, e.message, null)
        }
    }

    private fun handleUpdatePins(call: MethodCall, result: Result) {
        val pins = call.argument<List<String>>("pins")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "pins required", null)
        val hostname = call.argument<String>("hostname")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "hostname required", null)
        try {
            CShieldSSL.updatePins(pins = pins, hostname = hostname)
            result.success(null)
        } catch (e: IllegalStateException) {
            result.error(CShieldErrorCode.SSL_NOT_CONFIGURED, e.message, null)
        } catch (e: Exception) {
            result.error(CShieldErrorCode.NATIVE_ERROR, e.message, null)
        }
    }

    /**
     * Verifies a DER-encoded certificate (base64) using [CShieldSSL.getTrustManager],
     * which performs system CA validation + SPKI pin check.
     *
     * Flutter equivalent of CShieldTrustManager.checkServerTrusted() on Android.
     */
    private fun handleCheckServerTrusted(call: MethodCall, result: Result) {
        val certDerBase64 = call.argument<String>("certDer")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "certDer required", null)
        val host = call.argument<String>("host") ?: ""

        if (!CShieldSSL.isConfigured()) {
            return result.error(CShieldErrorCode.SSL_NOT_CONFIGURED, "CShieldSSL not configured", null)
        }

        try {
            val certBytes = Base64.decode(certDerBase64, Base64.DEFAULT)
            val factory = CertificateFactory.getInstance("X.509")
            val cert = factory.generateCertificate(ByteArrayInputStream(certBytes)) as X509Certificate

            // CShieldTrustManager.checkServerTrusted performs:
            //   1. System CA validation via X509TrustManagerExtensions (blocks user-installed CAs)
            //   2. SPKI pin verification against configured pins
            CShieldSSL.getTrustManager().checkServerTrusted(arrayOf(cert), "RSA")
            result.success(true)
        } catch (e: CertificateException) {
            result.success(false)
        } catch (e: Exception) {
            result.error(CShieldErrorCode.NATIVE_ERROR, e.message, null)
        }
    }

    /**
     * Performs an HTTPS request through OkHttp configured with
     * [CShieldSSL.getSSLSocketFactory] + [CShieldSSL.getTrustManager], so
     * certificate pinning runs at the TLS layer over the FULL chain.
     *
     * Flutter's Dart adapter (CShieldNativeHttpAdapter) delegates the pinned
     * host's requests here; this is what lets Flutter match intermediate/root
     * pins that pure-Dart networking cannot.
     */
    private fun handleHttpRequest(call: MethodCall, result: Result) {
        val method = (call.argument<String>("method") ?: "GET").uppercase()
        val url = call.argument<String>("url")
            ?: return result.error(CShieldErrorCode.INVALID_ARGUMENT, "url required", null)
        val headers = call.argument<Map<String, String>>("headers") ?: emptyMap()
        val body = call.argument<ByteArray>("body")
        val connectTimeoutMs = call.argument<Int>("connectTimeoutMs")
        val receiveTimeoutMs = call.argument<Int>("receiveTimeoutMs")
        val followRedirects = call.argument<Boolean>("followRedirects") ?: true

        if (!CShieldSSL.isConfigured()) {
            return result.error(CShieldErrorCode.SSL_NOT_CONFIGURED, "CShieldSSL not configured", null)
        }

        executor.execute {
            try {
                val builder = OkHttpClient.Builder()
                    .sslSocketFactory(CShieldSSL.getSSLSocketFactory(), CShieldSSL.getTrustManager())
                    .followRedirects(followRedirects)
                    .followSslRedirects(followRedirects)
                connectTimeoutMs?.let { if (it > 0) builder.connectTimeout(it.toLong(), TimeUnit.MILLISECONDS) }
                receiveTimeoutMs?.let { if (it > 0) builder.readTimeout(it.toLong(), TimeUnit.MILLISECONDS) }
                val client = builder.build()

                // OkHttp requires a body for POST/PUT/PATCH etc. and forbids one
                // for GET/HEAD. Content-Type is carried via the headers map below.
                val requestBody: RequestBody? = when {
                    body != null -> body.toRequestBody(null)
                    requiresRequestBody(method) -> ByteArray(0).toRequestBody(null)
                    else -> null
                }

                val reqBuilder = Request.Builder().url(url).method(method, requestBody)
                headers.forEach { (k, v) -> reqBuilder.header(k, v) }

                client.newCall(reqBuilder.build()).execute().use { response ->
                    val bodyBytes = response.body?.bytes() ?: ByteArray(0)
                    val headerMap = HashMap<String, List<String>>()
                    response.headers.names().forEach { name ->
                        headerMap[name] = response.headers.values(name)
                    }
                    val payload = hashMapOf<String, Any?>(
                        "statusCode" to response.code,
                        "reasonPhrase" to response.message,
                        "headers" to headerMap,
                        "body" to bodyBytes,
                    )
                    mainHandler.post { result.success(payload) }
                }
            } catch (e: javax.net.ssl.SSLPeerUnverifiedException) {
                mainHandler.post {
                    result.error(CShieldErrorCode.SSL_PIN_MISMATCH, e.message ?: "Certificate pin mismatch", null)
                }
            } catch (e: Exception) {
                mainHandler.post { result.error(CShieldErrorCode.NATIVE_ERROR, e.message, null) }
            }
        }
    }

    private fun requiresRequestBody(method: String): Boolean =
        method in setOf("POST", "PUT", "PATCH", "PROPPATCH", "REPORT")
}

File: example/ios/Classes/bridges/SslBridge.swift
import Flutter
import CShieldSDK
import Foundation

class SslBridge {

    // Cached when ssl.configure / ssl.updatePins is called so that
    // handleCheckServerTrusted can access pins without modifying the iOS SDK.
    private var cachedPins: [String] = []
    private var cachedHostname: String = ""

    func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "ssl.configure":
            guard let pins = args["pins"] as? [String],
                  let hostname = args["hostname"] as? String else { result(invalidArg()); return }
            cachedPins = pins
            cachedHostname = hostname
            CShieldSSL.configure(pins: pins, hostname: hostname)
            result(nil)
        case "ssl.updatePins":
            guard let pins = args["pins"] as? [String],
                  let hostname = args["hostname"] as? String else { result(invalidArg()); return }
            cachedPins = pins
            cachedHostname = hostname
            CShieldSSL.updatePins(pins: pins, hostname: hostname)
            result(nil)
        case "ssl.isConfigured":
            result(CShieldSSL.isConfigured())
        case "ssl.checkServerTrusted":
            handleCheckServerTrusted(args: args, result: result)
        case "ssl.httpRequest":
            handleHttpRequest(args: args, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - ssl.httpRequest

    /// Performs an HTTPS request through `CShieldSSL.urlSession()`, whose
    /// URLSession delegate (`CShieldTrustManager`) enforces certificate pinning
    /// at the TLS layer over the FULL chain.
    ///
    /// Flutter's Dart adapter (CShieldNativeHttpAdapter) delegates the pinned
    /// host's requests here so Flutter can match intermediate/root pins — which
    /// pure-Dart networking (leaf-only) cannot.
    private func handleHttpRequest(args: [String: Any], result: @escaping FlutterResult) {
        guard let urlString = args["url"] as? String, let url = URL(string: urlString) else {
            result(invalidArg()); return
        }
        guard CShieldSSL.isConfigured() else {
            result(FlutterError(code: CShieldErrorCode.sslNotConfigured,
                                message: "CShieldSSL not configured", details: nil))
            return
        }

        let method = (args["method"] as? String ?? "GET").uppercased()
        let headers = args["headers"] as? [String: String] ?? [:]
        let body = (args["body"] as? FlutterStandardTypedData)?.data
        let followRedirects = args["followRedirects"] as? Bool ?? true

        var request = URLRequest(url: url)
        request.httpMethod = method
        for (key, value) in headers { request.setValue(value, forHTTPHeaderField: key) }
        if let body = body { request.httpBody = body }
        if let receiveTimeoutMs = args["receiveTimeoutMs"] as? Int, receiveTimeoutMs > 0 {
            request.timeoutInterval = Double(receiveTimeoutMs) / 1000.0
        }

        // NOTE: URLSession follows redirects by default; `followRedirects == false`
        // is best-effort only (honoring it would require a task-level delegate).
        _ = followRedirects

        // MethodChannel results must be delivered on the main thread; URLSession
        // completions run on a background queue.
        func reply(_ value: Any?) { DispatchQueue.main.async { result(value) } }

        // Retain the session until completion by capturing it in the closure.
        let session = CShieldSSL.urlSession()
        let task = session.dataTask(with: request) { data, response, error in
            defer { session.finishTasksAndInvalidate() }

            if let error = error {
                let nsError = error as NSError
                // URLSession surfaces a pin/trust failure as a TLS cancel.
                let isTrustFailure = nsError.domain == NSURLErrorDomain
                    && (nsError.code == NSURLErrorCancelled
                        || nsError.code == NSURLErrorServerCertificateUntrusted
                        || nsError.code == NSURLErrorSecureConnectionFailed)
                reply(FlutterError(
                    code: isTrustFailure ? CShieldErrorCode.sslPinMismatch : CShieldErrorCode.nativeError,
                    message: error.localizedDescription, details: nil))
                return
            }
            guard let httpResponse = response as? HTTPURLResponse else {
                reply(FlutterError(code: CShieldErrorCode.nativeError,
                                   message: "Non-HTTP response", details: nil))
                return
            }

            var headerMap = [String: [String]]()
            for (key, value) in httpResponse.allHeaderFields {
                headerMap["\(key)"] = ["\(value)"]
            }
            let payload: [String: Any?] = [
                "statusCode": httpResponse.statusCode,
                "reasonPhrase": HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode),
                "headers": headerMap,
                "body": FlutterStandardTypedData(bytes: data ?? Data()),
            ]
            reply(payload)
        }
        task.resume()
    }

    // MARK: - ssl.checkServerTrusted

    /// Verifies a single DER-encoded leaf certificate (base64): SPKI pin check
    /// (via the native SDK's `CShieldSSL.spkiPin`) plus a structural SecTrust
    /// evaluation.
    ///
    /// NOTE: this path only receives the leaf cert from Dart, so it cannot do
    /// real CA chain validation or chain-based pin matching — the self-anchored
    /// SecTrust evaluation below is structural only. Full chain pinning + CA
    /// validation is enforced at the TLS layer by the native `CShieldSSL.urlSession()`
    /// delegate (`CShieldTrustManager`), which is the recommended integration path.
    private func handleCheckServerTrusted(args: [String: Any], result: @escaping FlutterResult) {
        guard let certDerBase64 = args["certDer"] as? String,
              let certData = Data(base64Encoded: certDerBase64, options: .ignoreUnknownCharacters),
              let cert = SecCertificateCreateWithData(nil, certData as CFData) else {
            result(invalidArg()); return
        }
        let host = args["host"] as? String ?? ""

        guard CShieldSSL.isConfigured() else {
            result(FlutterError(code: CShieldErrorCode.sslNotConfigured,
                                message: "CShieldSSL not configured",
                                details: nil))
            return
        }

        // Step 1: SPKI pin verification (does not require full chain).
        // Reuses the native SDK's single source of truth for pin computation
        // (CShieldTrustManager.computeSpkiPin) so the pin format never diverges
        // between the native URLSession delegate path and this Flutter bridge.
        let computed = CShieldSSL.spkiPin(for: cert)
        guard cachedPins.contains(computed) else {
            result(false); return
        }

        // Step 2: System CA validation with the leaf cert as its own anchor so
        // the evaluation succeeds even when the intermediate CA is not cached.
        // This mirrors Android's X509TrustManagerExtensions behaviour where the
        // system validates and we additionally enforce the SPKI pin above.
        let policy = SecPolicyCreateSSL(true, host as CFString)
        var trust: SecTrust?
        let certArray = [cert] as CFArray
        guard SecTrustCreateWithCertificates(certArray, policy, &trust) == errSecSuccess,
              let serverTrust = trust else {
            result(false); return
        }
        // Allow the cert to serve as its own anchor so CA chain validation
        // doesn't fail when only the leaf cert is provided.
        SecTrustSetAnchorCertificates(serverTrust, certArray)
        SecTrustSetAnchorCertificatesOnly(serverTrust, false)
        var cfError: CFError?
        let trusted = SecTrustEvaluateWithError(serverTrust, &cfError)
        result(trusted)
    }

    private func invalidArg() -> FlutterError {
        FlutterError(code: CShieldErrorCode.invalidArgument,
                     message: "Missing or invalid arguments",
                     details: nil)
    }
}
